### PR TITLE
feat(nbstore): new doc sync engine

### PR DESCRIPTION
**Pull request by EYHN**

### CODE REVIEW

**Comment by EYHN** [comment]
* **#9016** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9016?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#9110](https://github.com/toeverything/AFFiNE/pull/9110) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9110?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#9131](https://github.com/toeverything/AFFiNE/pull/9131) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9131?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#9084** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9084?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9070** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9070?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9037** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9037?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8996** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8996?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8918** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8918?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8918?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/toeverything?ref=merge-queue-instructions-in-stack-comment&prId=8918).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/common/nbstore/src/impls/idb/schema.ts [hunk 1]
```diff
@@ -94,6 +94,7 @@ export interface DocStorageSchema extends DBSchema {
       peer: string;
       docId: string;
       clock: Date;
+      pulledClock: Date;
```
Comment (forehalo) [inline packages/common/nbstore/src/impls/idb/schema.ts:97]: clock can represent pulled clock already?

File: packages/common/nbstore/src/sync/doc/peer.ts [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+          missing: newData,
+          state: serverStateVector,
+          timestamp: serverClock,
+        } = remoteDocRecord;
+        this.schedule({
+          type: 'save',
+          docId,
+          serverClock,
+        });
+        throwIfAborted(signal);
```
Comment (forehalo) [inline packages/common/nbstore/src/sync/doc/peer.ts:177]: why do we listen on abort again and again?

Comment (EYHN) [reply packages/common/nbstore/src/sync/doc/peer.ts:177]: To ensure that no further operations after the abort,
just add `throwIfAborted` after each await

File: packages/common/nbstore/src/impls/idb/schema.ts (hunk 1 above)
Comment (EYHN) [reply packages/common/nbstore/src/impls/idb/schema.ts:97]: For example, there are two docs on the remote clock:

{
a: 1,
b: 2
}

We will sync with priority, so it is possible that the one with a larger clock will be synchronized first. If b is sync first, we will store pulled clock:

{
b: 2
}

If the synchronization ends at this time, the next time we restart. We will use the maxClock = 2 we stored to get the update of the remote, and then we will miss a.

File: packages/common/nbstore/src/connection/connection.ts [hunk 3]
```diff
@@ -104,6 +104,25 @@ export abstract class Connection<T = any> {
...
+        if (status === 'connected') {
+          resolve();
+        }
+      });
+
+      signal?.addEventListener('abort', reason => {
+        reject(reason);
+      });
+    });
+  }
```
[the remark this answers is not in the document]
Comment (forehalo) [reply packages/common/nbstore/src/connection/connection.ts:124]: cleanup

**Review by forehalo — APPROVED** [review]

**Pull request merged**